esp32_ulp/preprocess.py

```diff
@@ -0,0 +1,53 @@
+from . import nocomment
+from .util import split_tokens
+
+
+class Preprocessor:
+    def __init__(self):
+        self._defines = {}
+
+    def parse_defines(self, content):
+        result = {}
+        for line in content.splitlines():
+            line = line.strip()
+            if line[:7] != "#define":
+                # skip lines not containing #define
+                continue
+            line = line[8:].strip()  # remove #define
+            parts = line.split(None, 1)
+            if len(parts) != 2:
+                # skip defines without value
+                continue
+            identifier, value = parts
+            tmp = identifier.split('(', 1)
+            if len(tmp) == 2:
+                # skip parameterised defines (macros)
+                continue
+            value = "".join(nocomment.remove_comments(value)).strip()
+            result[identifier] = value
+        self._defines = result
+        return result
+
+    def expand_defines(self, line):
+        found = True
+        while found:  # do as many passed as needed, until nothing was replaced anymore
+            found = False
+            tokens = split_tokens(line)
+            line = ""
+            for t in tokens:
+                lu = self._defines.get(t, t)
+                if lu != t:
+                    found = True
+                line += lu
+
+        return line
+
+    def preprocess(self, content):
+        self.parse_defines(content)
+        lines = [l for l in nocomment.remove_comments(content) if l != ""]
```

Comment:
when implementing the original code, i frequently and quickly ran into memory issues, so IIRC i called remove_comments very early to get rid of most of unneeded string content before doing anything else.

when testing this PR, it should be also tested on a standard esp32 device to see how its runtime behaviour regarding to memory usage / garbage collection is.

especially stuff working on the whole content string or splitting/joining all lines needs quite a bit of memory.
and we don't have gigabytes to use on the esp32...

Reply:
btw, if you kill empty lines, line numbers change - do we need them for userfriendly error messages?

Reply:
I managed to assemble quite a lot of examples without problem on a standard esp32, including with a database of defines (that is added to the preprocessor in a later commit) containing all ULP/RTC constants from the ESP-IDF.

That all worked for examples up to 6kb in input, thereafter I needed to add an explicit garbage collect - I added it into the `for line in lines` loop below, so that the garbage collector was called after every line. Worked fine but doesnt feel super elegant. I left it out of the pull-request for now, because I wanted to first get feedback on the general design of the solution and then to test/optimise more around memory usage (given that the current implementation managed all but 2 quite big examples I tried).

Reply:
are line numbers in error messages still correct if you remove empty lines?

Reply:
None of the errors show line numbers of the assembly source (always only the python call stacks), so this is not important. However, I guess empty lines don't take much to hold in memory, so we could keep them too. Maybe better if we ever do add error output related to the input code.

Reply:
yeah, i'ld expect the python interpreter to only use 1 empty string object and just refer to that from everywhere.

Reply:
oh, i thought we have line numbers in error messages.
didn't use this code for quite a while, sorry.

Reply:
I now preserve empty lines. Instead of "not using much more memory", it even resulted in 128 bytes **less** memory being used :) (independent of the size of the input text).